### PR TITLE
Video Module & JwPlayer video submodule : Initial Releases

### PR DESCRIPTION
**Pull request by karimMourra**
## Type of change

- [x] Feature

## Description of change
This PR is a Work in Progress and this branch is meant to be a feature branch. As components are implemented, PRs will be opened against this branch, allowing the community to review changes incrementally. 
The objective of this PR is to implement the Video Module proposed in https://github.com/prebid/Prebid.js/issues/6271 and detailed in the [official proposal doc](https://docs.google.com/document/d/1sueL8EXnD-DViTIEBNayAzNNrYocdHtmEl7433g0f5I/edit?usp=sharing).

- contact email of the adapter’s maintainer
 karim mourra - karim@jwplayer.com

## Other information

fulfills [proposal](https://docs.google.com/document/d/1sueL8EXnD-DViTIEBNayAzNNrYocdHtmEl7433g0f5I/edit?usp=sharing)
https://github.com/prebid/Prebid.js/issues/6271

### CODE REVIEW

**Comment by patmmccann** [comment]
@jimdigriz check this and https://github.com/jwplayer/Prebid.js/pull/5/files as an example of video player submodules to a video module. Curious for your feedback.

**Comment by jimdigriz** [comment]
@karimMourra sent me an older revision of the proposal that I commented on. This time around I now understand more about what it is trying to solve and it looks really useful. I am not a publisher so can only comment wearing an analytics/RTD hat.

I know the proposal states the scope as being 'video' here but as they are so close in implementation I think a statement about audio ads should be made, even if that statement is simply 'out of scope'. A description of what future work might look like and how it could build on the video module to support audio would be useful.

I do like the event hooks as they extend the existing events in an expected manner and are straight forward to use. They provide a technical solution for analytics purposes so it is no longer required to use [an 'ad server' module to create a VAST wrapper to crowbar in a VPAID shim](https://github.com/prebid/Prebid.js/pull/6309/commits/1e8595c25d10e10afc299f096af034e7ef1bfad2).

I do not really understand why `getOrtbParams()` exists. [Currently it resembles a 'capabilities' list](https://github.com/jwplayer/Prebid.js/pull/5/files#diff-b1754b14e9bd3351e66848cae18b894fd9d8517bc416fbd97e0f35dda68b2562R62-R126) and where it does not it mostly reflects the configuration passed to it (eg. `h`, `w`, ...) all of which should probably should be in the [`SETUP_COMPLETE` event](https://github.com/jwplayer/Prebid.js/pull/5/files#diff-b1754b14e9bd3351e66848cae18b894fd9d8517bc416fbd97e0f35dda68b2562R54) and then handled by VideoCore to set up the oRTB parameters there. It should also be described how the publisher could post-hook to fix up situations where a video player claims VAST 3 or later but does not support Ad Pods (eg. Teads, or they at least used to).

For all the VideoCore methods detailed in the proposal, instead of `divId` I would expect `AdSlot` which would be more consistent with other APIs found in Prebid.js.

Now for a bit of a larger thinking point, what would be interesting is if video sub-modules (actually all modules) were forced to make HTTP requests through Prebid.js:

1. publisher gains full control and enforcement (including cookie/tracking) of what is allowed
2. there is the option to provide hooks where another module can (optionally) rewrite the request and response.

Slightly off topic, now might be a good time to start looking to enforce this policy (starting in VideoCore) as there do not look to be many direct users (non-`ajax.js`):
```
alex@aineko:~/src/Prebid.js$ $ git grep -l -e sendBeacon -e 'new Image' -e 'XMLHttp' -e 'create.*img' -e 'fetch(' modules/
modules/adoceanBidAdapter.js
modules/adomikAnalyticsAdapter.js
modules/browsiRtdProvider.js
modules/craftBidAdapter.js
modules/mantisBidAdapter.js
modules/sharethroughAnalyticsAdapter.js
modules/smaatoBidAdapter.js
modules/sortableBidAdapter.js
modules/yuktamediaAnalyticsAdapter.js
```

Reworking `ajax.js` so that there is additionally a `fetch()`-esque API wrapper option would handle their needs of a non-blocking background/`` tracking. No need to add this functionality right now but forcing modules to push the request through Prebid.js means someone can add this functionality at a later date when they need it.

Back on topic, the ability to add a hook to rewrite the request/response (breakpointing on redirects) would make for a powerful tool.

One use case I can think of is if a publisher is trying to insert a verification partner. Ideally everything would support OMSDK/OMID (hah!) so the publisher could amend the VAST response to include further partners (if they are unable to make this change on their ad server) but when the publisher does not they can call out instead to a VAST/VPAID wrapper; of course if the video player says no to VPAID they can fallback further to event tracking pixels.

Another use case is to extract data from the `<Extensions>` block.

Though proposed, I do not actually see the need for a VAST parser, exposing the chain of VAST responses as XML documents makes for `querySelector()` being a really familiar way to interact with them. Maybe later Prebid.js may wish to provide a number of helper tools (I am thinking more `utils.js`/`querySelector()` orientated rather than `vast.js`) once the dust settles and code/function duplication becomes apparent.

**Comment by karimMourra** [comment]
Hi @jimdigriz thanks for the feedback.
`getOrtbParams()` will be called by the the module that bridges Prebid with Video-Core - let's call it Prebid-video. Prebid-Video will listen for Prebid hooks, say `before: bids requested`, call `getOrtbParams()` and write the result to the ad units, for the bid adapters to consume. This way the video params are provided by the Video Players, which are the subject matter experts, instead of the publisher's best guess. The reason all those params aren't passed with the `SETUP_COMPLETE` event is because they may change during the player's lice cycle, and in the case where a player is configured to play midrolls and postrolls, it's best to obtain the up to date oRTB params right before the auction, which may happen right before the ad's time cue. 

Audio Ads is a good idea, I remember it from our previous conversation, but I would like more feedback from the video player community before proceeding; the way jwplayer handles audio is detecting the item's media type and/or config options; i would rather see how the rest of the community handles audio before proposing a solution.

`AdSlot` vs `divId` is an interesting proposal. I think that should work. 

Re: http requests, are you talking about the requests to the ad server to get the ad creative? and the requests to the content servers to get the content ? Those are handled internally by the players; I don't think it would be feasible for the video players to request via Prebid

**Comment by jimdigriz** [comment]
> Re: http requests, are you talking about the requests to the ad server to get the ad creative? and the requests to the content servers to get the content ? Those are handled internally by the players; I don't think it would be feasible for the video players to request via Prebid

Hoping for as much as the requests as practical but you are right, outsourcing the actual media requests would be awkward.

Maybe an easier approach which involves no changes is to make law that all video players must support `URL.createObjectURL()` URLs so a publisher could do all the legwork involved in fetching the VAST tag (and making amendments if they wish) and then just pass in a resulting object URL using your existing proposed interface. Those not wishing to do that, can simply pass through the URL from the auction unchanged.

Our video module actually does this and wraps up the VAST tag with our changes and presents the resulting object URL: https://github.com/prebid/Prebid.js/pull/6309

**Comment by robertrmartinez** [comment]
Is this still a draft?

Assigning to @jsnellbaker for now so we do not lose track.

@karimMourra Please let us know when this is ready

**Comment by karimMourra** [comment]
Hi @robertrmartinez and @jsnellbaker: the modules that have been implemented are fully unit tested and can be merged.  They are not demoable or usable without the remaining modules. If you prefer, we can continue adding the remaining modules to this PR as they get implemented, or we can merge this PR and open separate PRs for the remaining modules as we go. The latter would result in smaller PRs that are easier to review instead of one large PR. I am open to either option.

**Comment by jsnellbaker** [comment]
Hi @karimMourra 

I personally would like to see more modules added here to that we can see a working demo.  If also it's possible, would you be able to add comments/descriptions to the various files to denote what their purpose and what they're meant to support in terms of the workflow?  If needed, please refer to the top of the `modules/userId/index.js` and its related files for what I mean.

Additionally, for the various 'constants' files, would it be possible to revise them to use a json structure; similar to what's used in the `src/constants.json` for core prebid.js?  It would simplify the import statements to perhaps a one/few parent variables, instead of the dozens of individual values.

Also could you clarify a bit on the 'storage' file and how it's meant to work with the other parts?  I tried to look through the google doc (maybe I missed it), but I didn't see a matching section.  I'm assuming it's meant to globally store a persistent session of certain values, but if so - would this be compatible with the multiple/parallel auctions that Prebid.js can run; I'm wondering if there's a case of overlapping/overwriting between different auction sessions (as each auction could run its own unique adunits).

Thanks!  This is looking like a good start.

**Comment by karimMourra** [comment]
Hi @jsnellbaker thank you for the feedback! I will add inline comments/descriptions and consolidate the constants . 

Not sure what you mean by the 'storage' file. Do you mean the `vendorDirectory` or `state.js` ? 
State.js is meant to be a simple state object that submodules can use to store information about the current video session. 
The Vendor Directory is meant to be populated at runtime by the submodules. Instead of having the core module import all the submodules, the submodule factories are added to the vendor directories. 

Regarding the docs, I agree that we could start writing them, but I would prefer to wait until the code is reviewed, in case any design changes become required. I hate writing docs and I would especially hate to rewrite them.

Lastly could I get your review on https://github.com/jwplayer/Prebid.js/pull/6/ ? The PR adds a layer connecting Prebid and the Video Module. It proposes a change in `events.js` to allow the Video events to be available without adding bloat to user who are using the module.

File: modules/.submodules.json [hunk 1]
```diff
@@ -52,5 +52,9 @@
   "fpdModule": [
     "enrichmentFpdModule",
     "validationFpdModule"
+  ],
+  "videoModule": [
+    "coreVideo",
+    "jwplayerVideoProvider"
```
Comment (patmmccann) [inline modules/.submodules.json:58]: does the gam ad server submodule need to be here?

File: src/events.js [hunk 2]
```diff
@@ -133,6 +133,10 @@ module.exports = (function () {
     return _handlers;
   };
 
+  _public.addEvents = function (events) {
+    allEvents.concat(events);
```
Comment (patmmccann) [inline src/events.js:137]: lgtm throwing an error on this, says ignores result https://lgtm.com/rules/1510745846115/

**Review by patmmccann — CHANGES_REQUESTED** [review]
add gam to submodules, fix lgtm alert

File: modules/jwplayerVideoProvider.js [hunk 3]
```diff
@@ -0,0 +1,997 @@
+import {
```
Comment (dgirardi) [inline modules/jwplayerVideoProvider.js:1]: These imports suffer from the same problem, but they should work because they are immutable. But if you look at the build output, you should see them duplicated, which is still not ideal.

**Review by dgirardi — COMMENTED** [review]
I got through only a portion of it, but I think this should not have worked in testing - can you merge master again so I can see how it did? (it's possible that what I talk about here behaves differently between prod and dev builds)

File: integrationExamples/videoModule/index.html [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+          console.log('adImpression: ', event);
+        });
+        pbjs.addAdUnits(adUnits);
+        pbjs.setConfig({
+          video: {
+            providers: [{
+              divId: 'vid-div-1',
+              vendorCode: 1,
+              playerConfig: {
+                licenseKey: "INSERT LICENSE KEY HERE",
```
[the remark this answers is not in the document]
Comment (karimMourra) [reply integrationExamples/videoModule/index.html:58]: Some players require providing a license key. The user does not need to publish his key in the PR. He would provide it in his local build for testing. We will be adding a license key to this PR for demo purposes. 
I a publisher's video player does not require a license key to be passed in, they can leave this blank. Alternatively, they can use their player's API to provide the license key; it is not required to pass it in via the video module's config; the video module's config is just for convenience.

[the remark this answers is not in the document]
Comment (dgirardi) [reply integrationExamples/videoModule/index.html:58]: anything that the end user needs to pass to `setConfig` is available to anyone who knows how to right click and inspect source. If these keys are something they pay for, it seems like something we should at least make clear and/or discourage.

[the remark this answers is not in the document]
Comment (karimMourra) [reply integrationExamples/videoModule/index.html:58]: The requirement of using a key is not up to us. If the player requires one, there's nothing we can do about it. This is just a convenience for the publisher to configure the player in one place. 
I agree with adding a warning in the docs and to let the reader know that if a safer alternative is available, that they should go with the latter.

File: modules/gamAdServerSubmodule.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  function getAdTagUrl(adUnit, baseAdTag) {
+    return dfp.buildVideoUrl({ adUnit: adUnit, url: baseAdTag });
+  }
+
+  return {
+    getAdTagUrl
+  }
+}
+function gamSubmoduleFactory() {
+  const dfp = $$PREBID_GLOBAL$$.adServers.dfp;
```
[the remark this answers is not in the document]
Comment (karimMourra) [reply modules/gamAdServerSubmodule.js:21]: out of curiosity, why so ? I've noticed that the macro is used in a lot of modules; what are the differences between the two?

File: modules/gamAdServerSubmodule.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import { GAM_VENDOR } from './videoModule/constants/vendorCodes.js';
+import { adServerDirectory } from './videoModule/vendorDirectory.js';
+
+/**
+ * @constructor
+ * @param {Object} dfpModule_ - the DFP ad server module
+ * @returns {AdServerProvider}
+ */
+function GamAdServerProvider(dfpModule_) {
```
[the remark this answers is not in the document]
Comment (karimMourra) [reply modules/gamAdServerSubmodule.js:9]: The intention is for this to be a general ad server interface. It should be agnostic to the video subsystem. WHy do you say it assumes the presence of the video module ? Is it because of the imports from the videoModule folder ?
I'm not sure I understand the suggestion of using a `getAdTagUrl` config for the video module. We would expect the publisher to provide the implementation for getting an ad tag ?

File: integrationExamples/videoModule/index.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (dgirardi) [reply integrationExamples/videoModule/index.html:58]: I know that many companies will automatically terminate service to api keys that have been leaked on the internet, so I'm not so sure that this would necessarily be a convenience - if it's a secret they need to protect, we should not allow them to make it public so easily. But again, I don't know if that's the case for these particular keys.

File: modules/gamAdServerSubmodule.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (dgirardi) [reply modules/gamAdServerSubmodule.js:21]: The difference is that `getGlobal` is normal javascript, while `$$PREBID_GLOBAL$$` is substituted by a custom babel plugin. Ultimately the first is just a wrapper around the second so it should not matter, but since our custom logic is probably more likely to break than the javascript VM, I think it's good practice to keep it constrained to the minimum possible.

File: modules/gamAdServerSubmodule.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (dgirardi) [reply modules/gamAdServerSubmodule.js:10]: what I'm thinking is if the only adserver-specific item you need now is `getAdTagUrl`, and GAM would work for most cases, we could start by making that the default so that we don't need to try to define a general adserver interface - that seems like a big, separate project.

[the remark this answers is not in the document]
Comment (karimMourra) [reply modules/gamAdServerSubmodule.js:10]: What do you mean GAM would work for most cases ? GAM only works if i'm using GAM as an ad server. Most people use GAM but it doesn't feel right to force or give preference to a specific vendor. AdLoox for example or Spring Serve might want to implement a submodule.

[the remark this answers is not in the document]
Comment (dgirardi) [reply modules/gamAdServerSubmodule.js:10]: Yes, I agree. What I'm saying is, rather than stubbing an ad server interface and providing only a GAM implementation as you do here, I would prefer to make it a video configuration option, with a default that works for GAM. Either way if you're not using GAM it will not work out of the box; but for the end user that is on another adserver, the config is easier to figure out - and we don't run the risk of making the wrong choices on the interface until we think about that explicitly and more generally. 

For example, I'm pretty sure a general ad server interface should be registered under `videModule/vendorDirectory`, because we'll definitely have situations where we need to understand the ad server but we don't have the video module installed. But I don't know yet where or how it should be registered, I think we can kick that can down the road and tackle it separately.

[the remark this answers is not in the document]
Comment (karimMourra) [reply modules/gamAdServerSubmodule.js:10]: @dgirardi I think I see your point. I agree that we can move the ad server components out of the videoModule folder; i'll address asap. 
As far as providing a video config option instead of an ad server module, I'm leaning towards keeping the ad server modules and modifying the ad server interface in the future. The goal is to make integration as easy as possible for the publisher; I think it's easier for them to name their ad server instead of coming up with the right implementation. 
Since the Video Module is the only client of the adServer interface, I'm fine with it changing as we get more input from ad servers and other potential clients.
To be clear, when configuring prebid and the video module, all the publisher has to provide is:
```
adServer: {
      vendorCode: "gam"
}
```
and optionally they can provide a `params` block to include any additional params that they wish to be used when the ad tag is generated. 
This is a lot simpler than the pub providing an implementation in a config.

File: modules/gamAdServerSubmodule.js [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import { GAM_VENDOR } from './videoModule/constants/vendorCodes.js';
+import { adServerDirectory } from './videoModule/vendorDirectory.js';
```
[the remark this answers is not in the document]
Comment (karimMourra) [reply modules/gamAdServerSubmodule.js:2]: @dgirardi I'm a bit confused by the need to use `hook.js` since this does work as expected. At runtime the directory is updated by all registered modules, and the parent module is able to call on submodules accordingly. 
Why is `hook.js` required ? I prefer avoiding it because it seems like unnecessary code; it relies on a 3rd party package which seems like overkill for something this simple

File: modules/jwplayerVideoProvider.js (hunk 3 above)
Comment (karimMourra) [reply modules/jwplayerVideoProvider.js:1]: Not sure I understand. What is wrong with 
```
import {
  PROTOCOLS, API_FRAMEWORKS, VIDEO_MIME_TYPE, PLAYBACK_METHODS, PLACEMENT, VPAID_MIME_TYPE
} from './videoModule/constants/ortb.js';
``` 
?

File: modules/gamAdServerSubmodule.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (dgirardi) [reply modules/gamAdServerSubmodule.js:2]: After pulling in the last update I did some digging and the reason it works is because you never actually read from `adServerDirectory` - if you did, you'd find it won't not contain what you expect. 

I'd like to give an example of what I mean (and why the rest of the codebase uses `hook.js`), but I am now confused as to the purpose of these registries. What do you intend them to be used for? can we sidestep the problem entirely by removing them for now? 

The general rule is modules should not import from other modules, only from core. The build process assumes that modules are independent; to simplify it, if you have two modules that do:

```javascript
// moduleA.js
export const someGlobalState = {};
export const someImmutableValue = 'foo';
```

```javascript
// moduleB.js
import {someGlobalState, someImmutableValue} from './moduleA.js';
someGlobalState[someImmutableValue] = 'bar';
```

they will get compiled to something like:

```javascript
symbols['moduleA'] = (function(){
    return {
       someGlobalState: {},
       someImmutableValue: 'foo'
    };
})();
symbols['moduleB'] = (function(){
   const {someGlobalState, someImmutableValue} = (function(){
    return {
       someGlobalState: {},
       someImmutableValue: 'foo'
    };
   })();
   moduleA[someImmutableValue] = 'bar';
   return {};
})();
```

that is, the build will not try to reuse the symbols exported from `moduleA`, because in general it cannot assume that it's  installed - you can choose to not include it in the package. Instead, it will re-compile it again and embed it in `moduleB`. If the symbols resolve to immutable values, this is just duplication - but if modules try to modify them, they will be each working on their copy, which will almost certainly be incorrect.

[the remark this answers is not in the document]
Comment (dgirardi) [reply modules/gamAdServerSubmodule.js:2]: Note that the above is the reason we have [this linter plugin](https://github.com/prebid/Prebid.js/blob/master/plugins/eslint/validateImports.js), which is supposed to error out when it detects cross-module imports. However, for reasons that I don't know, it allows cross-imports of anything except `index.js` if [the imported module is a directory](https://github.com/prebid/Prebid.js/blob/998e9343c472a1ed29037e7a6168a9771c5c13d4/plugins/eslint/validateImports.js#L25).

[the remark this answers is not in the document]
Comment (karimMourra) [reply modules/gamAdServerSubmodule.js:2]: The purpose of the `directories` is to serve as a registration process for submodules. When the prebid bundle is loaded on the page, the submodules included in the build will write to the shared directory. The `key` will be the submodule id string, and the `value` will be the submodule's factory function. When the parent module needs to instantiate a submodule, it can look up the factory. This way, a parent module does not need to import submodules; instead only the submodules included in the build will be linked to the parent module, by way of the `directory`

[the remark this answers is not in the document]
Comment (karimMourra) [reply modules/gamAdServerSubmodule.js:2]: the reason this works is because the `directory` is an `export const`

[the remark this answers is not in the document]
Comment (dgirardi) [reply modules/gamAdServerSubmodule.js:2]: no, the reason it works is because it doesn't do anything. At the moment the directories are useless, they are written to but never read.

[the remark this answers is not in the document]
Comment (karimMourra) [reply modules/gamAdServerSubmodule.js:2]: If it's useless then why does it work ? Have you tried the demo ? And actually the directory is being read
https://github.com/prebid/Prebid.js/pull/7264/files#diff-40c6bf504ebdf8d2f0211292ff3bc619bfca0448378f3bb0a0e658a463d37924R173

https://github.com/prebid/Prebid.js/pull/7264/files#diff-8f29ffeee22983e3a40ce2cb59a0deb6fb240a08bb8a993cc5fb50599a1f5664R75

[the remark this answers is not in the document]
Comment (dgirardi) [reply modules/gamAdServerSubmodule.js:2]: The demo, in the last version of this pr, throws an exception: 
> Uncaught TypeError: _videoModule_vendorDirectory_js__WEBPACK_IMPORTED_MODULE_2__.adServerDirectory is undefined

This is because [of this line](https://github.com/jwplayer/Prebid.js/blob/353930b49d316b39c1a44a16d860bce79ce65208/modules/gamAdServerSubmodule.js#L2).

But that's just an oversight, and not the real issue. If you fix that and try to read from the ad server directory *from the video module*, you will see that it is empty. The steps I took are:

- change that import to use `videoVendorDirectory` instead of `adServerDirectory` (I know it's incorrect, but it does not matter, beacause imports like these do not work)
- change the video config to use `adServer: { provider: 'GAM' }`, so that you actually trigger [this line](https://github.com/jwplayer/Prebid.js/blob/353930b49d316b39c1a44a16d860bce79ce65208/modules/videoModule/index.js#L110) to read from the directory
- put a breakpoint [here](https://github.com/jwplayer/Prebid.js/blob/353930b49d316b39c1a44a16d860bce79ce65208/modules/shared/parentModule.js#L40) and watch `submodules` be empty.

[the remark this answers is not in the document]
Comment (dgirardi) [reply modules/gamAdServerSubmodule.js:2]: Sorry, my example was not a good one - you want the correct directory to be imported for it to be convincing (`import { adServerDirectory } from './adServerModule/vendorDirectory.js';`). But this affects every use of that paradigm if you cross module boundaries:

no_submodules

[the remark this answers is not in the document]
Comment (karimMourra) [reply modules/gamAdServerSubmodule.js]: Ah I see, sorry I forgot to update the import statement in the gam submodule after moving the directory to the new adServer module folder. It should work now

[the remark this answers is not in the document]
Comment (dgirardi) [reply modules/gamAdServerSubmodule.js]: If you mean the demo, yes it does work, but the parentModule / directory pattern still doesn't - it's just not compatible with our module system. My suggestion is to either convert it to the module / submodule pattern defined in `hook.js`, or simplify this PR so that everything is contained in one module (at least to start; we can always extend later). The fact that the demo works shows that all these layers of abstractions are not needed right now - they are just being circumvented.

Also, I'm not sure if you see them on your end, but this PR has hundreds of LGTM alerts (https://lgtm.com/projects/g/prebid/Prebid.js/rev/pr-bf6d3f550989305375a32de43df4422b9c5c92b9), which makes it hard to follow - can you please address those?

[the remark this answers is not in the document]
Comment (dgirardi) [reply modules/gamAdServerSubmodule.js]: Yes regarding the LGTM alerts - by the way there's still one it's complaining about - unfortunately the bot is really keen on reminding us about them so it's drowning out real discussion in this thread.

Importing from modules results in incorrect code in the build output, as I tried to show above. All instances of `ParentModule` have an empty `submodules` because you cannot share state across modules this way. You would get the exact same behavior if you moved everything in `shared` and `adserver` inside the video module, because under the hood that's what webpack is doing. So that's at the minimum what's I'm asking you to do, because at least that way there's no extra dead code in the final bundle.

[the remark this answers is not in the document]
Comment (dgirardi) [reply modules/gamAdServerSubmodule.js]: Let me rephrase that - to a minimum, no module should `import` anything from any other module. It's unfortunate that our linter check happened to be blind to the way you are doing it here - I'm working on fixing that as well; I'm sorry you did all this work without knowing about this limitation.

File: integrationExamples/videoModule/index.html [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+                }
+              },
+              adServer: {
+                vendorCode: "gam"
+              }
+            }]
+          }
+        });
+
+        pbjs.onEvent('setupComplete', () => {
```
Comment (dgirardi) [inline integrationExamples/videoModule/index.html:84]: Should these events have 'video' in their name? they share the namespace with all of the rest of prebid.

File: modules/videoModule/index.js [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    if (!oRtbParams) {
+      return;
+    }
+
+    adUnit.mediaTypes.video = Object.assign({}, videoMediaType, oRtbParams.video);
+    let ortb2 = { ortb2: mergeDeep({}, getConfig('ortb2'), { site: oRtbParams.content }) };
+    pbGlobal.setConfig(ortb2);
+  }
+
+  function renderWinningBid(adUnit) {
```
Comment (dgirardi) [inline modules/videoModule/index.js:102]: I am not sure about this - tagging it for other reviewers. I am concerned that this takes away control from the publisher, as it seems very different from what happens now with the result of an auction (which is just returned for them to do what they wish with it). 

Giving them a function that can render a bid seems more in line with the current approach.

File: modules/videoModule/index.js [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        }, provider.divId);
+
+        const adServerConfig = provider.adServer;
+        if (adServerConfig) {
+          adServerCore.registerAdServer(adServerConfig);
+        }
+      });
+    });
+
+    requestBids.before(enrichAdUnits, 40);
```
Comment (dgirardi) [inline modules/videoModule/index.js:52]: is the explicit priority set to `40` necessary? if so, can you add a comment explaining why?

File: modules/shared/state.js [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+ * @function updateState
+ * @function getState
+ * @function clearState
+ */
+
+/**
+ * @summary factory to create a simple state object
+ * @returns {State}
+ */
+export default function stateFactory() {
```
Comment (dgirardi) [inline modules/shared/state.js:13]: this does not seem to add much value: why not use vanilla objects?

**Review by dgirardi — COMMENTED** [review]
There's an empty "constants.json" file in here - probably a leftover.

There's still a bunch of cross-module imports; I'll address those separately.

File: modules/jwplayerVideoProvider.js [hunk 12]
```diff
@@ -0,0 +1,997 @@
...
+  PROTOCOLS, API_FRAMEWORKS, VIDEO_MIME_TYPE, PLAYBACK_METHODS, PLACEMENT, VPAID_MIME_TYPE
+} from './videoModule/constants/ortb.js';
```
Comment (dgirardi) [inline modules/jwplayerVideoProvider.js:3]: The rule is still, "no cross-module imports". This is what gets built:

https://gist.github.com/dgirardi/3f9738dc3999ac3b4a433f299c10ccaf

All of these definitions are duplicated in every module that imports them. With the 'hook.js' pattern you adopted here you should have resolved the duplication of state, but code is still repeated. If we start adding more, distinct `videoProvider` modules that do the same imports, they would get repeated again.

It's a bit clunky, but I think the way to do what you want here is to remove `import`/`export`s and pass symbols you need through the `submodule` initializer - something like:

```
// videoModule/index.js
const sharedUtils = {...}
module('videoModule', function(submod) {
    registerSubmodule(submod(sharedUtils));
});
```

```
// someVideoProviderInstance.js
submodule('videoModule', function(sharedUtils) {
    return function submodFactory() {
            doStuffWith(sharedUtils);
    }
})

File: modules/jwplayerVideoProvider.js [hunk 13]
```diff
@@ -0,0 +1,997 @@
...
+import {
```
Comment (dgirardi) [inline modules/jwplayerVideoProvider.js:4]: The rule is still, "no cross-module imports". This is what gets built:

https://gist.github.com/dgirardi/3f9738dc3999ac3b4a433f299c10ccaf

All of these definitions are duplicated in every module that imports them. With the 'hook.js' pattern you adopted here you should have resolved the duplication of state, but code is still repeated. If we start adding more, distinct `videoProvider` modules that do the same imports, they would get repeated again.

It's a bit clunky, but I think the way to do what you want here is to remove `import`/`export`s and pass symbols you need through the `submodule` initializer - something like:

```
// videoModule/index.js
const sharedUtils = {...}
module('videoModule', function(submod) {
    registerSubmodule(submod(sharedUtils));
});
```

```
// someVideoProviderInstance.js
submodule('videoModule', function(sharedUtils) {
    return function submodFactory() {
            doStuffWith(sharedUtils);
    }
})

File: modules/adServerModule/adServer.js [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+  const parentModule = ParentModule(adServerSubmoduleBuilder);
+  const adServerCore = AdServerCore(parentModule);
+  return adServerCore;
+}
+
+function attachAdServerFactory(submoduleFactory) {
+  adServerDirectory[submoduleFactory.vendorCode] = submoduleFactory;
+}
+
+module('adServer', attachAdServerFactory)
```
Comment (dgirardi) [inline modules/adServerModule/adServer.js:87]: I still think we'd be better off delaying the adserver interface / directory and doing that separately; it would simplify this PR a lot.

At the moment, because there's no `index.js` in `adServerModule`, the dependency here is on `videoModule` (that's what `import`s the `module('adServer')` call). If you try to separate that out, you'll have the problem of trying to make this depend on `adServerModule`, which should depend on `shared`, which will be difficult to juggle without any `import`s. 

If you remove this, `shared` would only be used by `videoModule` and could be moved there which would allow you to keep most of the `import`s from it.

File: integrationExamples/videoModule/index.html (hunk 8 above)
Comment (karimMourra) [reply integrationExamples/videoModule/index.html:84]: I'm not against namespacing but the cost would be longer strings

File: modules/videoModule/index.js (hunk 9 above)
Comment (karimMourra) [reply modules/videoModule/index.js:102]: I understand that this goes against the pattern but the goal is to simplify the integration between prebid and video for publishers. I can bring this up next week at the Prebid Video sync to get the community's opinion.

File: modules/videoModule/index.js (hunk 10 above)
Comment (karimMourra) [reply modules/videoModule/index.js:52]: No reason, it seemed to be the common priority across modules. Not sure how to determine the right priority; any thoughts ?

Comment (dgirardi) [reply modules/videoModule/index.js:52]: Several ask for priority 40 because they need to run after GDPR processing, which is set at priority 50. In your case I don't think it matters - but I haven't tested.

File: modules/videoModule/index.js (hunk 9 above)
Comment (dgirardi) [reply modules/videoModule/index.js:102]: I agree with discussing it in the PMC. I have nothing against easy integration but I can also imagine publishers that can't use this because it has no way to accept their logic on which bids to render.

File: modules/videoModule/index.js (hunk 10 above)
Comment (karimMourra) [reply modules/videoModule/index.js:52]: I understand. The only requirement is for the function to run before **buildRequests** is called on the Bid Adapters, since the objective of the function is to write all of the ortb video and content info, for the bid adapters to consume before making a bid.

File: modules/shared/state.js (hunk 11 above)
Comment (karimMourra) [reply modules/shared/state.js:13]: makes the code a bit DRYer. Useful for future video submodules since they will need to track state as well.

File: modules/adServerModule/adServer.js (hunk 14 above)
Comment (dgirardi) [reply modules/adServerModule/adServer.js:87]: I'll expand on this a bit - a complete ad server interface, when we get there, will almost certainly need a lot of support from core. There are many ugly instances where core has special cases for GAM that would need to be captured in it (see https://github.com/prebid/Prebid.js/issues/7638).

So most likely the future complete version of your `adServerModule` would need to be included in core, with specific ad server submodules filling in the blanks - it's unlikely that much the work you'd have to do here to keep it out of core (and removing all the `import`s) could be reused. 

But at the same time, I don't think the problem is well defined enough yet to justify asking you to make this more future-proof. Hence why I think it's better to sidestep this completely for now.

File: modules/shared/state.js (hunk 11 above)
Comment (dgirardi) [reply modules/shared/state.js:13]: I disagree somewhat - the abstraction is so thin, I don't think `state.updateState(newState)` is appreciably better than `Object.assign(state, newState)`. It wouldn't be worse either, except that extra code has a cost in the browser. 

But I don't mind it enough to argue any more than this :)

File: integrationExamples/videoModule/index.html (hunk 8 above)
Comment (dgirardi) [reply integrationExamples/videoModule/index.html:84]: since in the code you have a separation between video and prebid events, you could just prepend 'video' (or similar) when you propagate from one to the other - the string would be longer but the code hardly would. 

This may be another topic to bring up in the PMC; once these names are set we're stuck with them.

Comment (karimMourra) [reply integrationExamples/videoModule/index.html:84]: What is PMC ?

File: modules/videoModule/index.js (hunk 9 above)
Comment (karimMourra) [reply modules/videoModule/index.js:118]: addressed; `renderWinningBid` is only called if there is no bidsBackHandler and a render function is exposed

File: modules/jwplayerVideoProvider.js (hunk 13 above)
Comment (karimMourra) [reply modules/jwplayerVideoProvider.js]: @dgirardi, this has been addressed

File: modules/jwplayerVideoProvider.js [hunk 15]
```diff
@@ -0,0 +1,1060 @@
...
```
[the remark this answers is not in the document]
Comment (karimMourra) [reply modules/jwplayerVideoProvider.js:1]: Hi @dgirardi I've tried a way to solve this by supporting dependencies in modules. COuld you please take a look at https://github.com/jwplayer/Prebid.js/pull/27 for a POC ?

**Comment by dgirardi** [comment]
@karimMourra what do you think of reopening a new PR to cut down on the length of this thread?

**Comment by karimMourra** [comment]
@dgirardi the video.js submodule is incomplete; I will try to add the missing work and reopen! thanks!
